### PR TITLE
Add Hooks

### PR DESCRIPTION
**Pull request by dpordomingo**

### CODE REVIEW

File: behaviours/common.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
+package behaviours
+
```
Comment (erizocosmico) [inline behaviours/common.go]: move all the package to the root, we don't want to go importing a bazillion packages for very core functionality

File: behaviours/identifiable.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package behaviours
+
+import "github.com/src-d/go-kallax/common"
+
+//Identifiable must be implemented by those values that can be identified by an ID
+type Identifiable interface {
+	// GetID returns the ID.
+	GetID() common.ID
```
Comment (erizocosmico) [inline behaviours/identifiable.go]: `ID` should be in root, not in `common` package

File: behaviours/identifiable.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	// SetID overrides the ID.
+	SetID(id common.ID)
+	// Identify sets the ID if it does not exist and returns the ID of the object
+	Identify() common.ID
```
Comment (erizocosmico) [inline behaviours/identifiable.go]: why do you need this?

File: behaviours/identifiable.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+// Identificator modelates an object that knows about its ID and generates it if neccessary
+type Identificator struct {
```
Comment (erizocosmico) [inline behaviours/identifiable.go]: why do you need this?

File: behaviours/timestampable.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+import "time"
+
+// Timestampable must be implemented by those object that can be timestamped
+type Timestampable interface {
+	Timestamp()
+}
+
+// Timestamp modelates an object that knows about when was created and updated
+type TimestampDates struct {
```
Comment (erizocosmico) [inline behaviours/timestampable.go]: i'd name this `Timestamps` or `Timestampable`

File: behaviours/timestampable.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
+package behaviours
+
+import "time"
+
+// Timestampable must be implemented by those object that can be timestamped
+type Timestampable interface {
```
Comment (erizocosmico) [inline behaviours/timestampable.go]: why do you need this interface?

File: common/id.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
+package common
```
Comment (erizocosmico) [inline common/id.go]: move to root

File: model.go [hunk 8]
```diff
@@ -1,63 +1,38 @@
...
-
-func (id ID) IsEmpty() bool {
-	return uuid.UUID(id) == uuid.Nil
-}
-
+// Model is the base type of the items that are stored
 type Model struct {
-	ID        ID
+	behaviours.Identificator
+	behaviours.TimestampDates
```
Comment (erizocosmico) [inline model.go]: timestampable should not be included by default

File: model.go [hunk 9]
```diff
@@ -1,63 +1,38 @@
...
-}
-
-func (id ID) IsEmpty() bool {
-	return uuid.UUID(id) == uuid.Nil
-}
-
+// Model is the base type of the items that are stored
 type Model struct {
-	ID        ID
+	behaviours.Identificator
```
Comment (erizocosmico) [inline model.go]: put `ID` field here, it's not necessary to embed anything here

File: model.go [hunk 10]
```diff
@@ -1,63 +1,38 @@
...
-}
-
+// IsPersisted returns whether this Model is new in the store or not.
 func (m *Model) IsPersisted() bool {
 	return m.persisted
 }
 
-func (m *Model) setPersisted(isPersisted bool) {
-	m.persisted = isPersisted
+func (m *Model) setPersisted() {
```
Comment (erizocosmico) [inline model.go]: if `setPersisted` is not really a `set` method I would name it `persist` or `persisted`

File: model.go [hunk 11]
```diff
@@ -66,11 +41,33 @@ func (m *Model) setWritable(w bool) {
...
+type Persistable interface {
+	// IsPersisted returns whether this Model is new in the store or not.
+	IsPersisted() bool
+	setPersisted()
+}
+
+// Writable must be implemented by those values that defines internally
+//  if they can be sent back to the database to be stored with its changes.
+type Writable interface {
+	IsWritable() bool
```
Comment (erizocosmico) [inline model.go]: should have comment

File: store.go [hunk 12]
```diff
@@ -53,9 +53,7 @@ func (s *Store) Insert(record Record) error {
 		return ErrNonNewDocument
 	}
 
-	if record.GetID().IsEmpty() {
-		record.SetID(NewID())
-	}
+	record.Identify()
```
Comment (erizocosmico) [inline store.go]: I would keep it as it was. It's not going to be used anywhere else and feels a little pointless to keep this method just for this ad-hoc logic

File: behaviours/common.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+// Hookable must be implemented by those entities that attend to all DB Hooks
```
Comment (erizocosmico) [inline behaviours/common.go]: why do you need the hookable interface? The thing about events is that you should be able to implement just the ones you need. This interface forces you to implement all of them

File: behaviours/common.go [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+	BeforeUpdater
+	AfterUpdater
+	BeforeInserter
+	AfterInserter
+	BeforeDeleter
+	AfterDeleter
+}
+
+// AfterLoader must be implemented by those entities listening the AfterLoad event
+type AfterLoader interface {
```
Comment (erizocosmico) [inline behaviours/common.go]: i don't know if we are going to need these interfaces at all. The code will be generated and we will not check if a record fulfills them, so the only use is if someone maybe wants to check it by hand

File: model.go (hunk 10 above)
Comment (Serabe) [inline model.go]: `persisted` does not sound ok for a method that does something

Comment (dpordomingo) [inline model.go]: `person.Persist()` should save the person into the db, what is not the purpose of this method.

**Comment by dpordomingo** [comment]
@erizocosmico the proposed changes was applied

File: model.go (hunk 10 above)
Comment (erizocosmico) [inline model.go]: ok, fair enough :+1:

File: model.go [hunk 15]
```diff
@@ -41,6 +42,25 @@ func (m *Model) setWritable(w bool) {
 	m.writable = w
 }
 
+// GetID returns the ID.
+func (i *Model) GetID() ID {
+	return i.ID
+}
+
+// SetID overrides the ID.
+//  The ID should not be modified once it has been set and stored in the DB
```
Comment (erizocosmico) [inline model.go]: extra space before the "The"

File: model.go [hunk 16]
```diff
@@ -41,6 +42,25 @@ func (m *Model) setWritable(w bool) {
...
+func (i *Model) SetID(id ID) {
+	i.ID = id
+}
+
+//Identifiable must be implemented by those values that can be identified by an ID
```
Comment (erizocosmico) [inline model.go]: missing space after `//`

File: model.go [hunk 17]
```diff
@@ -51,24 +71,26 @@ type Persistable interface {
...
 	setWritable(bool)
 }
 
 // ColumnAddresser must be implemented by those values that exposes its properties
 //  under pointers, identified by its property names
 type ColumnAddresser interface {
-	// ColumnAddress returns a pointer to the object property identified by the passed string
-	//  or an error if that property does not exist
+	// ColumnAddress returns a pointer to the object property identified by the
+	// passed string or an error if that property does not exist
```
Comment (erizocosmico) [inline model.go]: i'd say "column name" instead of "passed string"

File: resultset.go [hunk 18]
```diff
@@ -10,6 +10,10 @@ import (
 // not allowed.
 var ErrRawScan = fmt.Errorf("result set comes from raw query, use RawScan instead")
 
+// ErrStop is an error returned by a callback to stop the iteration in a
```
Comment (erizocosmico) [inline resultset.go]: you can remove this one from here, I added it somewhere else on the codegen branch (or I can delete mine and move this one to `store.go`)

File: model.go (hunk 10 above)
Comment (Serabe) [inline model.go]: `setAsPersisted`? Wasn't we naming that method `Save`?

File: resultset.go (hunk 18 above)
Comment (dpordomingo) [inline resultset.go]: Removed

**Pull request merged**